File: typescript/src/booking/application/integration/payment/handlers/handlePaymentResultFromPayment.ts
import type { PaymentResultIntegrationEvent } from '@/payment/integration/paymentResultIntegrationEvent.js';
import { PaymentResult } from '@/booking/application/integration/payment/adapter/paymentResult.js';
import { BookingConfirming } from '@/booking/application/usecases/bookingConfirming.js';
import { BookingRejecting } from '@/booking/application/usecases/bookingRejecting.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class HandlePaymentResultFromPayment {
  private readonly _paymentResult: PaymentResult;
  private readonly _confirming: BookingConfirming;
  private readonly _rejecting: BookingRejecting;

  constructor(paymentResult: PaymentResult, confirming: BookingConfirming, rejecting: BookingRejecting) {
    requireDependency(paymentResult, 'paymentResult');
    this._paymentResult = paymentResult;
    requireDependency(confirming, 'confirming');
    this._confirming = confirming;
    requireDependency(rejecting, 'rejecting');
    this._rejecting = rejecting;
  }

  handle(event: PaymentResultIntegrationEvent): void {
    const command = this._paymentResult.adapt(event);
    if (command === null) {
      return;
    }

    switch (command.kind) {
      case 'ConfirmBooking':
        this._confirming.invoke(command);
        break;
      case 'RejectBooking':
        this._rejecting.invoke(command);
        break;
    }
  }
}

File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/integration/payment/handlers/HandlePaymentResultFromPayment.java
package it.giannibombelli.wsc2026.booking.application.integration.payment.handlers;

import it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands;
import it.giannibombelli.wsc2026.booking.application.integration.payment.adapter.PaymentResult;
import it.giannibombelli.wsc2026.booking.application.usecases.BookingConfirming;
import it.giannibombelli.wsc2026.booking.application.usecases.BookingRejecting;
import it.giannibombelli.wsc2026.common.utils.Require;
import it.giannibombelli.wsc2026.payment.integration.PaymentResultIntegrationEvent;

/**
 * Handler cross-BC che reagisce agli esiti di pagamento pubblicati da {@code payment}
 * confermando o rifiutando la prenotazione corrispondente.
 */
public final class HandlePaymentResultFromPayment {

    private final PaymentResult paymentResult;
    private final BookingConfirming confirming;
    private final BookingRejecting rejecting;

    public HandlePaymentResultFromPayment(PaymentResult paymentResult,
                                          BookingConfirming confirming,
                                          BookingRejecting rejecting) {
        Require.requireDependency(paymentResult, "paymentResult");
        Require.requireDependency(confirming, "confirming");
        Require.requireDependency(rejecting, "rejecting");
        this.paymentResult = paymentResult;
        this.confirming = confirming;
        this.rejecting = rejecting;
    }

    public void handle(PaymentResultIntegrationEvent event) {
        BookingConfirmationCommands cmd = paymentResult.adapt(event);
        if (cmd == null) {
            return;
        }
        switch (cmd) {
            case BookingConfirmationCommands.ConfirmBooking confirmBooking -> confirming.invoke(confirmBooking);
            case BookingConfirmationCommands.RejectBooking rejectBooking -> rejecting.invoke(rejectBooking);
        }
    }
}
